src/core/forteinstance.cpp

```diff
@@ -0,0 +1,49 @@
+/*******************************************************************************
+ * Copyright (c) 2024 Primetals Technologies Austria GmbH, Martin Erich Jobst
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *    Alois Zoitl, Martin Jobst - initial implementation and rework communication
+ *                                infrastructure
+ *******************************************************************************/
+
+#include "forteinstance.h"
+#include "forte_device_config.h"
+
+
+std::unique_ptr<CDevice> C4diacFORTEInstance::createDev(const std::string &paMGRID) {
+  return FORTE_DEVICE::createDev(paMGRID);
```

Comment:
I tend to disagree. Implementing a correct RMT_RES just with a constructor is not as easy as it looks. Having a factory method like proposed by  @mx990  is much easier and gives more flexibility.

Reply:
but you wouldn't touch the RMT_RES. `RMT_DEV` would take `paMGRID` and call `setMGR_ID(paMGRID);` directly in the constructor, storing it locally in its `DI`. `RMT_RES` is constructed as before.
Calling `RMT_DEV::initialize()` could be done probably in the constructor, but since it could fail and to avoid `execptions` , you could have it in `C4diacFORTEInstance::startupNewDevice` before starting the device.

Reply:
> `RMT_DEV` would take `paMGRID` and call `setMGR_ID(paMGRID);` directly in the constructor, storing it locally in its `DI`.

The input variable `MGR_ID` of `RMT_RES` has not yet been initialized in the constructor of `RMT_DEV`. This is done later in `RMT_DEV::initialize()`, which delegates to `RMT_RES::initialize()`. As a result, we cannot call `RMT_DEV::setMGR_ID(...)` in the constructor of `RMT_DEV`.

> Calling `RMT_DEV::initialize()` could be done probably in the constructor

Please refer to https://isocpp.org/wiki/faq/strange-inheritance#calling-virtuals-from-ctors for the reason why calling `initialize()` (or any other virtual function, directly or indirectly) in the constructor is problematic.

Reply:
@cochicde I tried your proposed approach. In general to code is much nicer. However I can not manage to shut-down 4diac FORTE with it. While with the last commit everything works nice and fine.

Reply:
I tried it here and seems to be working fine https://github.com/azoitl/4diac-forte/pull/2 storing the string locally since the interface is not ready at that point as mentioned. 

I let the call to `initialize` still outside of it for now.

is there a reason the interface is not setup in the constructor?

Reply:
yes, I also tried the reworking `C4diacFORTEInstance` but could not go further. I think this class looks good as you have it and wouldn't complicate more. 

in https://github.com/azoitl/4diac-forte/pull/2 I left only the changes related to removing the `createDev` method in each derive class from `CDevice`

Reply:
If you don't mind I would take your PR as inspiration but when @mx990 PR #120 is merged it can be implemented a bit better. So I would wait for that and rework mine.

Reply:
@cochicde @mx990 I think now it is a nice solution. What do you think?

Reply:
looks good!